### DISCUSSION TITLE
Why should I want Rust in my project?

### DISCUSSION THREAD

**Discussion by jari-kuipers** [tauri-apps/tauri#9990] [General]
Hi everybody,

So I just caught wind of Tauri 2.0 which is coming up.

I have to say, I'm very excited and I think the tech that's being developed here is amazing and beautiful.
As a Flutter and Android native dev, I already see this as a way better framework for running web as an app than Electron, React Native and (obviously) .NET MAUI.

**The question that I'm about to ask comes from a place of pure interest and I mean no disprespect in any way when asking:**
Why do I still have to do Rust?

Let's say I'm working on my web stack and I need to hook into the platform. Let's say Android because I'm an Android dev, right? I'd register my Kotlin plugin in Rust and from there on I can communicate between JS and Kotlin. Great!

There are Rust files in my project though. All the talks that I see seem to put "calling Rust from JS" in the spotlight. The Rust files are always there and actually they're kind of cluttering because I now have files for yet another programming language in my project.

I get a feeling that I'm supposed to think that this is nice and a big benefit from the talks that I see and the docs that I'm reading. But I can't see it.

Why should I want Rust in my project?

**Comment by JonasKruckenberg** [comment]
First off: Hello :) Welcome!

To answer your question: Rust is core to Tauris architecture, alls its OS bindings and internal code are written in it and all of the team is thoroughly "rust pilled" as they say ;) 

Tauri is designed to leverage the fact that there is a compile-step to your application: That means contrary to the traditional design electron, nodejs and even browsers use wehere you have a pre compiled runtime that just runs plaintext JS files (or renders them) Tauri actually bakes them into the binary at compile time, it also strips out all unused code from the runtime and performs a number of optimizations.
All those things require the presence of a compiler and Tauris architecture uses the Rust compiler for that. 

So in fact Tauri as a "runtime" doesn't really exist: it is just a Rust library that you can pull in to do things. To make this more approachable we then offer the JS API and getting started templates so you can be productive without having to touch the Rust files. But Rust remains core to the tool.

One of the benefits of Tauris architecture is that you can easily pull in native code too though, Rust being the easiest ofc but integrating with C, C++, Swift, Objective-C etc is also possible.
We advertise this as a feature (and Rust in particular) because the ability to call native code from JS is incredibly powerful (as you mentioned). 
Rust is just what we as the Tauri authors believe to be a very very good language for writing this native code in, and in fact if you can and or have the need for it I would encourage you to try Rust! Tauri tends to be a very approachable way of learning the language.

**Reply by jari-kuipers** [reply]
Glad to be here. Thank you!

And thank you for your comprehensive reply and I wholeheartedly agree! Rust is an amazing language and I love using it even though I don't get many opportunities to write a lot of code in it.

In my personal day-to-day I don't run into many use cases where I would benefit from having a programming language with the computational capabilities of Rust at my disposal. Why, in Flutter for example I use Dart to develop the bulk of my application and hook into Kotlin and/or Swift whenever I need to access native capabilities.

And that is exactly why I wondered why Rust files are so present in Tauri projects. Since the use cases are so limited in my case, I would probably never need or want to touch an rs-file that's in my project even though I love the language.

Do I understand correctly that the reason why Rust is so present in a Tauri project even though the developer might not need Rust to make their cross-platform application is because Tauri is a Rust library with web capabilities and not the other way around? Therefore saying that I'd want a Tauri app to not contain Rust files because I don't need them would be like saying I want to make a Rust project without Rust.

Is that line-of-thinking about right?

**Reply by JonasKruckenberg** [reply]
> Tauri is a Rust library with web capabilities and not the other way around? 

Exactly! It's a rust library that lets you open windows with arbitrary web content (and more) in them plus a few bits and bobs.

The default templates just come "preconfigured" with everything already hooked up.

**Reply by rulatir** [reply]
> The default templates just come "preconfigured" with everything already hooked up.

Except they don't. When writing one's first few projects, one will be recompiling Rust code dozens of times (and each time is a coffee break) upon realizing that this plugin or that permission is also needed. Eventually a reusable template settles that really has everything in place. The ones that come "preconfigured" are VERY far from that.

**Reply by FabianLars** [reply]
So you'd rather want a bloated template with all the plugins etc where you end up having to remove everything if you want it to be safe and efficient? Sounds more like something a community template should do as it doesn't really fit our current philosophy.

**Reply by lazarok09** [reply]
all comunity templates sucks (on frontend perspective) by adding more and more along the way. I guess you can keep it that way @FabianLars 100%
